zebROS_ws/src/behaviors/src/2023_align_to_grid.cpp

```diff
@@ -0,0 +1,280 @@
+      g.y = grid->second[1];
+      gridLocations_[std::stoi(grid->first.substr(3))] = g;
+      ROS_INFO_STREAM(std::stoi(grid->first.substr(3)) << " " << g.y << " " << g.piece);
+    }
+
+    // need to load grid stations here.
+
+    as_.start();
+  }
+
+  ~AlignToGridAction(void)
+  {
+  }
+
+  void callback(const field_obj::DetectionConstPtr& msg) {
+    // need to use transforms...
+    // oh wait no we don't! we can tell spline srv what frame id it is relevant to
+    latest_ = *msg;
+  }
+
+  std::optional<int> findClosestApriltag(const field_obj::Detection &detection) {
+    double minDistance = std::numeric_limits<double>::max();
+    int closestTag;
+    if (detection.objects.size() == 0) {
+      return std::nullopt;
+    }
+    for (field_obj::Object obj : detection.objects) {
+      double d = distance(obj.location);
+      if (d < minDistance) {
+        minDistance = d;
+        closestTag = std::stoi(obj.id);
+      }
+    }
+    return std::optional<int>{closestTag};
+  }
+
+  int getAllianceRelativeStationNumber(int alliance, int gridStation) {
+    // 1 = red alliance, 0 = blue
+    if (alliance == 1) {
+      return (8-(gridStation-1))+1;
+    }
+    return gridStation;
+  }
+
+  void executeCB(const behavior_actions::AlignToGrid2023GoalConstPtr &goal)
+  {
+    // TODO: do we want to find the tag closest to *us* or closest to the grid?
+    // right now it finds the one closest to the robot
+    // I'm not sure what is most accurate
+    auto closestTag = findClosestApriltag(latest_);
+    if (closestTag == std::nullopt) {
+      as_.setAborted(result_); // no tag found
+      ROS_ERROR_STREAM("[2023_align_to_grid] No AprilTags found :(");
+      return;
+    }
+    int closestId = closestTag.value();
+    Tag tag = tags_[closestId];
+
+    int gridStation = getAllianceRelativeStationNumber(goal->alliance, goal->grid_id);
+    GridLocation gridLocation = gridLocations_[gridStation];
+
+
+    // 1 is at the left, 9 is at the right
+
+    // e.g. the tag we find is at (1,1) -- src and we want to go to (2,2) -- dest
+    // need dest - src
+
+    geometry_msgs::Point offset;
+    offset.x = -1.0;
+    if (goal->alliance == 1) {
+      offset.y = gridLocation.y - tag.location.y; // should be gridLocation.y - tag.location.y
+    } else {
+      offset.y = -(gridLocation.y - tag.location.y);
+    }
+    offset.z = 0;
+    ROS_INFO_STREAM("ydif: " << gridLocation.y - tag.location.y);
+
+    geometry_msgs::Pose pose;
+    pose.position = offset;
+
+    tf2::Quaternion q;
+    q.setRPY(0, 0, 0); // facing the tag
+    geometry_msgs::Quaternion qMsg = tf2::toMsg(q);
+
+    pose.orientation = qMsg;
+    // set orientation to quaternion with yaw = tagRotation, use setRPY
+
+    behavior_actions::PathToAprilTagGoal aprilGoal;
+    aprilGoal.id = closestId;
+    aprilGoal.tagRotation = tag.rotation;
+    aprilGoal.offset = pose;
+
+    client_.sendGoal(aprilGoal);
+
+    ros::Rate r(100);
```

Comment:
Can we get away with a lower rate here?

Reply:
i'm going to change it to 30 Hz so we move a maximum of 10cm between when we preempt and when the robot actually stops (assuming we are going 3 meters per second), is that ok?